client_integration_test.go

```diff
@@ -1249,6 +1249,57 @@ func TestClientReadSequential(t *testing.T) {
 	}
 }
 
+type lastChunkErrSequentialWriter struct {
+	expected      int
+	written       int
+	writtenReturn int
+}
+
+func (w *lastChunkErrSequentialWriter) Write(b []byte) (int, error) {
+	chunkSize := len(b)
+	w.written += chunkSize
+	if w.written == w.expected {
+		return w.writtenReturn, errors.New("test error")
+	}
+	return chunkSize, nil
+}
+
+func TestClientWriteSequential_WriterErr(t *testing.T) {
+	sftp, cmd := testClient(t, READONLY, NODELAY)
+	defer cmd.Wait()
+	defer sftp.Close()
+
+	d, err := ioutil.TempDir("", "sftptest-writesequential-writeerr")
+	require.NoError(t, err)
+
+	defer os.RemoveAll(d)
+
+	var (
+		content    = []byte("hello world")
+		shortWrite = 2
+	)
+	w := lastChunkErrSequentialWriter{
+		expected:      len(content),
+		writtenReturn: shortWrite,
+	}
+
+	f, err := ioutil.TempFile(d, "write-sequential-writeerr-test")
+	require.NoError(t, err)
+	fname := f.Name()
+	n, err := f.Write(content)
+	require.NoError(t, err)
+	require.Equal(t, n, len(content))
+	require.NoError(t, f.Close())
+
+	sftpFile, err := sftp.Open(fname)
+	require.NoError(t, err)
+	defer sftpFile.Close()
+
+	gotWritten, gotErr := sftpFile.writeToSequential(&w)
+	require.NotErrorIs(t, io.EOF, gotErr)
+	require.Equal(t, int64(shortWrite), gotWritten)
```

Comment:
The API is confusing on this matter:
[`func ErrorIs(t TestingT, err error, target error, msgAndArgs ...interface{})`](https://pkg.go.dev/github.com/stretchr/testify/require#ErrorIs),
but:
[`func Equal(t TestingT, expected interface{}, actual interface{}, msgAndArgs ...interface{})`](https://pkg.go.dev/github.com/stretchr/testify/require#Equal)

This is one of the central reasons why I tell people to avoid using `require` and `assert`. Getting the arguments in the wrong order is way too easy to do, and without checking the documentation, who knows which one is which. (Especially, since it’s `require.ErrorIs(t, actual, expect)` but `require.Equals(t, expect, actual)` )

This is especially true, when one is coming from a language that incentivizes writing equality tests as `constValue == testValue` so that mistakenly only using one `=` will not produce an unexpectedly valid assignment, but an invalid unassignable left-hand value. In Go, you cannot put an assignment in a condition, and so there is no way to accidentally use a single `=` assignment like this. This idiom is probably the very reason why `require.Equal` orders its parameters in that order… [an order that is discouraged in Go](https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures).

Instead, something like this (without the comments which are here for study, rather than for good code practice):
```
if got != int64(shortWrite) {
	// In the `require.Equals`, the two arguments are cast into `interface{}`,
	// which means you lose the same compile-time type safety you would normally have.
	// It’s way better to use the code the exact same as your caller would use the code,
	// with verbose `if err != nil { }` et al. included.
	
	// Note the order here, `if got != expected { … }`, and the message orders `got` then `expected`.
	t.Errorf("sftpFile.Write() = %d, but expected %d", got, shortWrite)
}
```

Also, we don’t want to `require` either of these tests anyways. We only want to `assert` it in this case. (Since both `written` and `err` have valid values, testing both of them every test makes sense. `require` is only appropriate when the test is a requirement for the remaining tests to have any meaningful result and/or not panic.